proposals/NNNN-max-records-per-node.md

```diff
@@ -0,0 +1,114 @@
+* Author(s): [Anupama Chandrasekhar](https://github.com/anupamachandra), [Mike Apodaca](https://github.com/mapodaca-nv)
+* Sponsor: TBD
+* Status: **Under Consideration**
+
+# [MaxRecordsPerNode(count)] Attribute for NodeOutputArray
+
+## Introduction
+
+This specification describes the HLSL and DXIL details for a new [NodeArrayOutput](https://microsoft.github.io/DirectX-Specs/d3d/WorkGraphs.html#node-output-attributes) attribute `[MaxRecordsPerNode(count)]` that specifies the maximum number of records that can be output to a specific node in a node output array. See the [MaxRecordsPerNode]() specifications for more details.
+
+## Motivation
+
+For `NodeArrayOutput`, the node output attribute `[MaxRecords(count)]` specifies the maximum number of records that can
+be output across the entire node array.  This attribute alone is insufficient for determining how records are
+distributed across an output array.  For example, consider an output node array specification of
+`[MaxRecords(N)][NodeArraySize(N)]`. All N records could be sent to one node in the array, or one record could be
+sent to each of the N nodes in the array, or the records could be spread in an arbitrary fashion across multiple nodes
+in the array.  An implementation cannot distinguish these different use cases.
+
+When determining backing store memory requirements, an implementation must assume the worst-case of `MaxRecords` written
+to any single node in the output array.  However, a common use-case is for a small number records to be written to
+select nodes in a very large array of nodes.  Some implementations can take advantage of this knowledge to significantly
+reduce the backing store memory requirements while maintaining peak performance.
+
+## Proposed solution
+
+We propose a new node output attribute called `MaxRecordsPerNode`. This parameter is only required for output node
+arrays.  This attribute specifies the maximum number of records that can be written to any single output node within a
+node array.
+
+## Detailed design
+
+### HLSL Additions
+
+Add a new node output attribute:
+
+| Attribute | Required | Description |
+|:---       |:--------:|:------------|
+| `[MaxRecordsPerNode(count)]` | Y | For `NodeArrayOutput`, specifies the maximum number of records that can be output to a node within the array.  Exceeding this results in undefined behavior.  This attribute can be overridden via the `NumOutputOverrides / pOutputOverrides` option when constructing a work graph.  This attribute has no impact on existing node output limits. |
+
+This attribute will be required starting with a future Shader Model version.
+Since this may cause compilation failures with existing Work Graphs, a new compiler command line option will be
+introduced to replace the compiler error with a warning and implicitly set the value of `MaxRecordsPerNode`
+equal to `MaxRecords`. (TODO: Specify the command line option)
```

Comment:
The way to do this is to use a warning that defaults to error (a `DefaultError` warning) with a warning group that can be addressed at the command line to override this default (change to warning or silence).

```suggestion
Since this may cause compilation failures with existing Work Graphs, this will
be a `DefaultError` warning assigned to a warning group named
`hlsl-require-max-records-per-node` to allow a command-line override.
The value of `MaxRecordsPerNode` will be set equal to `MaxRecords`.
```